### ISSUE TITLE
"Socket timed out, no connections available"

### ISSUE THREAD

**Issue by saltthehash** [esdalmaijer/EyeTribe-Toolbox-for-Matlab#4]
Before I even begin, I should probably start off by mentioning two things: This is my first time using python and I am running this on OS X Yosemite. I used the python source code to compile the server using py2app. For some reason, it can't find connect to the socket. As per the instructions, I run the EyeTribe server, and then I run the .app toolbox server. It opens, then stops responding for the designated 2 minute timeout, then ends with an error. I saw that the error in the console was from line 41 of EyeTribe_Matlab_server.py which calls for the error "Socket timed out, no connections available". I tried using different ports then recompiling, I tried using my computer's local name as a string, I even tried defining MLIP as MLIP = socket.gethostname(). Do you have any ideas of why this might be happening or how  could possibly fix it? Thank you!

**Comment by esdalmaijer** [comment]
Thanks for trying this on OS X!

As far as I know, the code should work on other platforms than Windows. I'm not familiar with py2app, so I couldn't tell you if that's the culprit. The error suggests a timeout, which could mean that a connection is actually opened, but then not used.

Just to be sure, you should be doing the following:
1) Connect the EyeTribe to a USB 3.0 port
2) Start the EyeTribe Server
3) Start the .app you've created from the Python source
**4) Start your Matlab script**

It's quite crucial to to step 4 within two minutes of step 3, as the Python server will simply look for an available connection for two minutes, before giving up. The connection it's looking for, is a socket created in the Matlab `eyetribe_init` function.

If this doesn't work, you could try using a different port number (it could be that 5666 is already occupied in your system, although normally this would result in a different error). Make sure to change it in the [Python source](https://github.com/esdalmaijer/EyeTribe-Toolbox-for-Matlab/blob/master/python_source/EyeTribe_Matlab_server.py#L15), as well as in [eyetribe_init.m](https://github.com/esdalmaijer/EyeTribe-Toolbox-for-Matlab/blob/master/EyeTribe_for_Matlab/eyetribe_init.m#L18).

Please do let me know if you experience any issues after trying these suggestions, or if these solved your issue.

Good luck!

**Comment by shandelman116** [comment]
Ha, I feel so stupid! It was indeed step 4 that I forgot to do!!! Just got it working with the matlab script you provided! Thank you so much!

By the way, py2app was designed to be very similar to py2exe. I was able to figure that part out after reading up on the documentation. The code I used to compile was slightly different than that for the py2exe (apparently it was more similar to py2exe in an older version of py2app). I actually generated it using a python function that comes with py2app called py2applet, but had to tweak the file so that it would include pytribe.py. I can send you the setup.py file and the command in terminal to compile it with py2app. I may as well also post the code here for now as well:

setup.py:

    from setuptools import setup

    APP = ['EyeTribe_Matlab_server.py']
    DATA_FILES = ['pytribe.py']
    OPTIONS = {'argv_emulation': True}

    setup(
        app=APP,
        data_files=DATA_FILES,
        options={'py2app': OPTIONS},
        setup_requires=['py2app'],
    )

Then use the following commands in terminal to compile:

    cd /path/to/pythonfiles/
    python setup.py py2app

More information about py2app can be found here: https://pythonhosted.org/py2app/

Thank you so much Edwin!!!

**Comment by saltthehash** [comment]
Ha, I feel so stupid! It was indeed step 4 that I forgot to do!!! Just got it working with the matlab script you provided! Thank you so much!

By the way, py2app was designed to be very similar to py2exe. I was able to figure that part out after reading up on the documentation. The code I used to compile was slightly different than that for the py2exe (apparently it was more similar to py2exe in an older version of py2app). I actually generated it using a python function that comes with py2app called py2applet, but had to tweak the file so that it would include pytribe.py. 

However, I just learned that since OS X actually comes with Python, the py2app compilation isn't even necessary. Instead, the server can simply be run by going to Terminal in OS X, setting the working directory to the folder with the python script and simply running it like this (changing the directory to the one where the python files are located) so that the script can just run within Terminal instead of as a standalone app:

```
$ cd /path/to/python_source/
$ python EyeTribe_Matlab_server.py
```

This was quite an interesting learning experience for me! Although I unnecessarily compiled it into a .app format, it was still interesting to do since this was my first experience with Python and I got to learn a little bit more about it. It seems like it might be useful to have a shell script that will run both the EyeTribe server and the Matlab server in one swoop.

Again, thank you so much Edwin for your help! I truly appreciate you making this toolbox!
